content/python/get_virtual_guest_usage_records.md

```diff
@@ -0,0 +1,94 @@
+---
+title: "CPU and Memory usage in Virtual servers"
+description: "How the portal page retrieves the memory and cpu usage"
+date: "2019-04-10"
+classes:
+    - "SoftLayer_Virtual_Guest"
+    - "SoftLayer_Metric_Tracking_Object"
+tags:
+    - "metric"
+    - "metrics"
+    - "virtual server"
+    - "virtualguest"
+---
+```py
+import SoftLayer
+
+class Virtual:
+    def __init__(self):
+        client = SoftLayer.create_client_from_env()
+        self.virtual_service = client['Virtual_Guest']
+        self.metric_service = client['Metric_Tracking_Object']
+
+    def _get_metric_tracking_object(self, guest_id):
+        return self.virtual_service.getMetricTrackingObjectId(id=guest_id)
+
+    def _calculate_averages(self, records):
+        # total usage and how many records per item type
+        total = {}
+        for item in records:
+            counter, date_time, item_type = item.values()
+            total[item_type] = [x + y for x, y in zip(total.get(item_type, [0, 0]), [counter, 1])]
+
+        # average of each item_type
+        averages = {}
+        for item, value in total.items():
+            average = value[0] / value[1]
+            averages[item] = average
+
+        return averages
+
+    def get_cpu_usage(self, guest_id, start_date, end_date):
+        metric_object_id = self._get_metric_tracking_object(guest_id)
+
+        # build the SoftLayer_Container_Metric_Data_Type array
+        valid_types = []
+        # set range according number of cpu items you want to retrieve the metric data
+        for i in range(16):
```

Comment:
couldn't you get this number `16` from the virtual guest object itself?

Reply:
that is a very good idea, I will retrieve the `startCpus` using masks